go/vt/vtctl/topo.go

```diff
@@ -175,6 +182,52 @@ func copyFileToTopo(ctx context.Context, ts *topo.Server, cell, from, to string)
 	return err
 }
 
+func commandTopoPost(ctx context.Context, wr *wrangler.Wrangler, subFlags *flag.FlagSet, args []string) error {
+	cell := subFlags.String("cell", topo.GlobalCell, "topology cell to use for the write. Defaults to global cell.")
+
+	subFlags.Parse(args)
+
+	if subFlags.NArg() != 3 {
+		return fmt.Errorf("TopoPost: needs object type name, json blob and destination")
+	}
+
+	// Currently only supports topodatapb.Shard but this is only the first cut
+	objectName := subFlags.Arg(0)
+
+	// Path in topo cell that we want to update
+	destination := subFlags.Arg(1)
+
+	// Json encoded protobuf. More than likely the output of a single message retrieved from
+	// topocat
+	jsonStr := subFlags.Arg(2)
+
+	// TODO: make this type resolution a lot friendlier
+	if strings.ToLower(objectName) != strings.ToLower(topo.ShardFile) {
+		return fmt.Errorf("Only 'shard' datatype supported")
+	}
+
+	var shard topodatapb.Shard
+	err := jsonpb.UnmarshalString(jsonStr, &shard)
+	if err != nil {
+		return fmt.Errorf("couldn't unmarshal json: %v", err)
+	}
+
+	data, err := proto.Marshal(&shard)
```

Comment:
TIOLI: a free idea

```go
type TopoPostMarshaler interface {
    Unmarshal(string) error
    Marshal() ([]byte, error)
}

type struct shardTopo struct {
    shard topodatapb.Shard
}

func (st *shardTopo) Unmarshal(s string) error {
    err := jsonpb.UnmarshalString(s, &st.shard)
    if err != nil {
       return fmt.Errorf("couldn't unmarshal json: %v", err)
    }
    return nil
}

func (st *shardTopo) Marshal() ([]byte, error) {
    data, err := proto.Marshal(&st.shard)
    if err != nil {
        return nil, fmt.Errorf("Error marshaling protobuf to bytes: %v", err)
    }
    return data, nil
}

// and then the resolution becomes:
marshalers := map[string]TopoPostMarshaler{
    "shard": &shardTopo{},
}

marshaler, ok := marshalers[strings.ToLower(objectName)]
if !ok {
    return fmt.Errorf("Unsupported object type: %v", objectName)
}

if err := marshaler.Unmarshal(jsonStr); err != nil {
    return err
}
data, err := marshalers.Marshal()
```

Reply:
requisite: this is obviously not threadsafe, if you wanted that you'd need to switch to a marshaler factory... but that seems unnecessary for a vtctl-slack command

Reply:
ooh. For my understanding, why isn't this thread safe? is ti because of the map access?

Reply:
nah, because the map contains a single instance of the marshaler -- so if you grab the marshaler from different threads and Unmarshal->Marshal you'll be stopming on the contents of the `shardTopo` struct.

```go
marshalerFactory := map[string]func()TopoPostMarshaler{
    "shard": func() TopoPostMarshaler { return &shardTopo{} },
}
```

would make it such that you could make `marshalerFactory` shared around as much as you want since it doesn't bake any state into the map contents.

Reply:
I had a better idea when we were chatting this morning:

```go
type TopoPostProcessor interface {
    Process(string) ([]byte, error)
}

type struct shardTopo struct {}

func (_ shardTopo) Process(s string) ([]byte, error) {
    var shard topodatapb.Shard
    err := jsonpb.UnmarshalString(s, &shard)
    if err != nil {
       return nil, fmt.Errorf("couldn't unmarshal json: %v", err)
    }

    data, err := proto.Marshal(&shard)
    if err != nil {
        return nil, fmt.Errorf("Error marshaling protobuf to bytes: %v", err)
    }
    return data, nil
}

// and then the resolution becomes:
handlers := map[string]TopoPostProcessors{
    "shard": shardTopo{},
}

handler, ok := handlers[strings.ToLower(objectName)]
if !ok {
    return fmt.Errorf("Unsupported object type: %v", objectName)
}

data, err := handler.Process(jsonStr)
if err != nil {
    return err
}
// now use data
```